x-pack/plugins/fleet/server/services/epm/elasticsearch/template/template.ts

```diff
@@ -486,41 +462,73 @@ const getDataStreams = async (
     dataStreamName: dataStream.name,
     replicated: dataStream.replicated,
     indexTemplate,
+    composedOfTemplates,
   }));
 };
 
+const getTemplateWithNameEndingIn = (composedOfTemplates: TemplateMap, suffix: string) => {
+  const mappingComponentTemplateName =
+    Object.keys(composedOfTemplates).find((templateName) => templateName.endsWith(suffix)) || '';
+
+  return composedOfTemplates[mappingComponentTemplateName];
+};
+
+const rolloverDataStream = (dataStreamName: string, esClient: ElasticsearchClient) => {
+  try {
+    // Do no wrap rollovers in retryTransientEsErrors since it is not idempotent
+    const path = `/${dataStreamName}/_rollover`;
+    return esClient.transport.request({
+      method: 'POST',
+      path,
+    });
+  } catch (error) {
+    throw new Error(`cannot rollover data stream ${dataStreamName} - ${error}`);
+  }
+};
+
 const updateAllDataStreams = async (
   indexNameWithTemplates: CurrentDataStream[],
   esClient: ElasticsearchClient,
   logger: Logger
 ): Promise<void> => {
-  const updatedataStreamPromises = indexNameWithTemplates.map(
-    ({ dataStreamName, indexTemplate }) => {
-      return updateExistingDataStream({ dataStreamName, esClient, logger, indexTemplate });
-    }
-  );
+  const updatedataStreamPromises = indexNameWithTemplates.map((templateEntry) => {
+    return updateExistingDataStream({ esClient, logger, ...templateEntry });
+  });
   await Promise.all(updatedataStreamPromises);
 };
 const updateExistingDataStream = async ({
   dataStreamName,
   esClient,
   logger,
   indexTemplate,
+  composedOfTemplates,
 }: {
   dataStreamName: string;
   esClient: ElasticsearchClient;
   logger: Logger;
   indexTemplate: IndexTemplate;
+  composedOfTemplates: TemplateMap;
 }) => {
-  const { settings, mappings } = indexTemplate.template;
-
-  // for now, remove from object so as not to update stream or data stream properties of the index until type and name
-  // are added in https://github.com/elastic/kibana/issues/66551.  namespace value we will continue
-  // to skip updating and assume the value in the index mapping is correct
-  delete mappings.properties.stream;
-  delete mappings.properties.data_stream;
-  // try to update the mappings first
+  const mappingComponentTemplate = getTemplateWithNameEndingIn(
+    composedOfTemplates,
+    MAPPINGS_TEMPLATE_SUFFIX
+  );
+  // @ts-expect-error 2339
+  const mappings = mappingComponentTemplate?.template?.mappings;
+
   try {
+    if (!mappingComponentTemplate || !mappings?.properties) {
+      // if there is no mapping component template then something has
+      // gone wrong, so rollover the stream to get the new mapping.
+      await rolloverDataStream(dataStreamName, esClient);
+      return;
+    }
```

Comment:
I'm not sure I'm following the need for this change. How would rolling over the data stream resolve a problem with us not being able to generate the correct templates in `getTemplateWithNameEndingIn`?

If we do need this, we should also have a unit test that sets up the preconditions for this explicitly to better explain the need for this behavior.

Reply:
haven't pushed the refactor yet